### PR TITLE
Deprecate and replace IHostingEnvronment & IApplicationLifetime

### PR DESCRIPTION
**Pull request by Tratcher**
#966
- IHostingEnvronment -> IHostEnvronment 
- IApplicationLifetime -> IAppLifetime

@davidfowl since the old types still exist the namespace conflict remains. It might be better to remove the old types.

@glennc

### CODE REVIEW

File: src/Hosting/Abstractions/src/IAppLifetime.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading;
+
+namespace Microsoft.Extensions.Hosting
+{
+    /// <summary>
+    /// Allows consumers to perform cleanup during a graceful shutdown.
+    /// </summary>
+    public interface IAppLifetime
```
Comment (davidfowl) [inline src/Hosting/Abstractions/src/IAppLifetime.cs]: Call it `IHostingApplicationLifetime` I hate the abbreviated App and it's anti the .NET naming guidelines

**Comment by JunTaoLuo** [comment]
😕 I would also be confused about why there are two types that look similar. Is this a case where we need to share code between Extensions and AspNetCore better?

File: src/Hosting/Abstractions/src/IAppLifetime.cs (hunk 1 above)
Comment (Tratcher) [reply src/Hosting/Abstractions/src/IAppLifetime.cs]: IHostingApplicationLifetime's not great either. @glennc any ideas?

**Comment by Tratcher** [comment]
@JunTaoLuo We we'd started with generic host then web host could have shared more types. As is we're breaking the thing most people haven't used yet rather than the one everyone is already using.

File: src/Hosting/Abstractions/src/IAppLifetime.cs (hunk 1 above)
Comment (glennc) [reply src/Hosting/Abstractions/src/IAppLifetime.cs]: Why is this interface duplicated again? It's identical right?

Comment (Tratcher) [reply src/Hosting/Abstractions/src/IAppLifetime.cs]: So the name no longer conflicts with the one in Microsoft.AspNetCore.Hosting.

**Comment by davidfowl** [comment]
This really is a soft breaking change as the old types still exist and work.

File: src/Hosting/Abstractions/src/IAppLifetime.cs (hunk 1 above)
Comment (Tratcher) [reply src/Hosting/Abstractions/src/IAppLifetime.cs]: How about ``IHostApplicationLifetime` so it at least mirrors the other new one and isn't quite as long?

Comment (davidfowl) [reply src/Hosting/Abstractions/src/IAppLifetime.cs]: YUP.

**Comment by Tratcher** [comment]
@davidfowl @glennc how about replacing IHostingEnvironment in the templates as well?
https://github.com/aspnet/AspNetCore/blob/1c6651a1ae54371de34358df1622a1fa645ef0cf/src/ProjectTemplates/Web.ProjectTemplates/content/EmptyWeb-CSharp/Startup.cs#L21

**Comment by davidfowl** [comment]
@Tratcher I thought about that but it might be wrong because it doesn't have the WebRoot or WebRootFileProvider.

**Comment by Tratcher** [comment]
Fair point, I'll leave that one.

File: src/Hosting/Abstractions/src/IAppLifetime.cs (hunk 1 above)
Comment (glennc) [reply src/Hosting/Abstractions/src/IAppLifetime.cs]: HostApplicationLifetime is what I expected but I don't remember why this has to be renamed and not moved. Do we not now have two identical things in two namespaces?

**Comment by davidfowl** [comment]
cc @jdom @ReubenBond

**Comment by ReubenBond** [comment]
Thanks for the heads up, @davidfowl. Orleans is not using `IHostingEnvironment` or `IApplicationLifetime` so this should not affect us.

**Comment by PinpointTownes** [comment]
I'm confused by this change too. So `IHostEnvironment` will replace `IHostingEnvironment` (and will be the new generic host interface) whereas `IHostingEnvironment` will be ASP.NET Core-specific and expose things like the web root file provider?

If so, wouldn't renaming ASP's `IHostingEnvironment` to `IWebHostingEnvironment` be way less confusing?

**Comment by davidfowl** [comment]
Yes but too breaking, that’s why we chose this approach. Nobody is using the generic host APIs as yet but everyone that has used ASP.NET Core is using IHostingEnvironment. It’s in the default template.

**Comment by Tratcher** [comment]
@PinpointTownes we talked it over today and updated the proposal: https://github.com/aspnet/Extensions/issues/966#issuecomment-463752915

**Comment by PinpointTownes** [comment]
> Nobody is using the generic host APIs as yet but everyone that has used ASP.NET Core is using IHostingEnvironment.

It likely doesn't count, but I've used it a lot in `IHostedService`/`BackgroundService` implementations 😅 

> @PinpointTownes we talked it over today and updated the proposal: #966 (comment)

Nice, it's less confusing this way. That said, wouldn't marking ASP.NET Core's `IHostingEnvironment`/`IHostApplicationLifetime` as obsolete and removing them in 4.0 be simpler than introducing a third type?

The result will be roughly the same in both cases, anyway, since `Microsoft.AspNetCore.Hosting.IHostingEnvironment` will be marked as obsolete. Migrating to `Microsoft.Extensions.Hosting.IHostingEnvironment` or `Microsoft.Extensions.Hosting.IHostingEnvironment` doesn't seem that different 😄

**Comment by Tratcher** [comment]
@PinpointTownes marking one obsolete isn't sufficient, you can't use either type. Having AspNetCore.Hosting.IHostingEnvironment and Extensions.Hosting.IHostingEnvironment causes compiler errors in your app if you include both namespaces (which is becoming more common).

**Comment by PinpointTownes** [comment]
Makes sense. It's just sad you have to introduce a third type to mitigate this situation, that exists since 2.2 (or 2.1?) 😄

**Comment by Tratcher** [comment]
Ready for review.

I also fixed startvs.cmd so it's not hardcoded to to the root sln.

**Comment by davidfowl** [comment]
What is codecheck?

**Comment by Tratcher** [comment]
Codecheck verifies things like the reference assemblies are up to date. I need to regen them with these changes.

File: src/Hosting/Abstractions/ref/Directory.Build.props [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<Project>
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory)..\..\..\..\, Directory.Build.props))\Directory.Build.props" />
+
+  <PropertyGroup>
+    <NoWarn>CS0618</NoWarn>
```
Comment (pakrym) [inline src/Hosting/Abstractions/ref/Directory.Build.props]: You don't need this, right?

**Comment by hoborm** [comment]
> 
> 
> #966
> 
>     * IHostingEnvronment -> IHostEnvronment
> 
>     * IApplicationLifetime -> IAppLifetime
> 
> 
> @davidfowl since the old types still exist the namespace conflict remains. It might be better to remove the old types.
> 
> @glennc

There are some typos in this thread missing the "I" in IHostEnv**I**ronment. Might be important when people want to search for the new interfaces in the repo and not seeing this issue.

**Comment by Tratcher** [comment]
Posted https://github.com/aspnet/AspNetCore/issues/7749 for discoverability.

Typos corrected.

**Pull request merged**